OMGShop/Models/Paginator.swift

```
@@ -0,0 +1,77 @@
+//
+//  Paginator.swift
+//  OMGShop
+//
+//  Created by Mederic Petit on 5/3/18.
+//  Copyright © 2018 Mederic Petit. All rights reserved.
+//
+
+import OmiseGO
+
+enum LoadingState {
+    case idle
+    case loading
+}
+
+class Paginator<T: Decodable> {
+
+    var page: Int = 1
+    var perPage: Int = 10
+    var state: LoadingState = .idle
+    var reachedLastPage: Bool = false
+
+    var successClosure: ObjectClosure<[T]>?
+    var failureClosure: FailureClosure?
+
+    var currentRequest: OMGRequest<OMGJSONPaginatedListResponse<T>>?
+
+    init(page: Int,
+         perPage: Int,
+         successClosure: ObjectClosure<[T]>?,
+         failureClosure: FailureClosure?) {
+        self.page = page
+        self.perPage = perPage
+        self.successClosure = successClosure
+        self.failureClosure = failureClosure
+    }
+
+    func reset() {
+        self.page = 1
+        self.cancelLoading()
+        self.reachedLastPage = false
+    }
+
+    func cancelLoading() {
+        self.currentRequest?.cancel()
+        self.state = .idle
+    }
+
+    func loadNext() {
+        guard self.state == .idle && !self.reachedLastPage else { return }
+        self.state = .loading
+        self.load()
+    }
+
+    func load() {
+        // Override in subclass
```

Shouldn't this raise an exception instead of a simple warning?